### PR TITLE
[Gemma4] Add MoE Calibration Support

### PR DESCRIPTION
**Pull request by dsikka**
SUMMARY:
- Add MoE calibration support
- Add e2e FP8 and NVFp4 examples
- Requires; https://github.com/vllm-project/vllm/pull/39045 for vLLM serving

Testing:
- Generated Checkpoints:
1. https://huggingface.co/RedHatAI/gemma-4-26B-A4B-it-FP8-Dynamic
2. https://huggingface.co/RedHatAI/gemma-4-26B-A4B-it-NVFP4

## Summary by CodeRabbit

* **New Features**
  * Added per-expert unpacking and calibration support for Gemma4 MoE models to enable expert-level calibration and quantization workflows.
  * Exported tooling to operate on Gemma4 expert modules and include them in calibration flows.

* **Documentation**
  * Added example scripts demonstrating NVFP4 (W4A4) and FP8 dynamic (W8A8) quantization flows, dataset preprocessing, custom batching for calibration, and saving quantized models and processors.

### CODE REVIEW

**Comment by dsikka** [comment]
@coderabbitai review this PR

File: examples/quantization_w8a8_fp8/gemma4_example.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    ],
+)
+
+# Apply quantization.
+oneshot(model=model, recipe=recipe)
+
+# Save to disk in compressed-tensors format.
+SAVE_DIR = "/raid/engine/dsikka/" + MODEL_ID.rstrip("/").split("/")[-1] + "-FP8-Dynamic"
+model.save_pretrained(SAVE_DIR, save_compressed=True)
+processor.save_pretrained(SAVE_DIR)
```
[the remark this answers is not in the document]
Comment (dsikka) [reply examples/quantization_w8a8_fp8/gemma4_example.py]: @coderabbitai
"/raid/engine/dsikka/" remove this

File: examples/quantization_w8a8_fp8/gemma4_example.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        "re:.*router",
+        "re:.*vision_tower.*",
+    ],
+)
+
+# Apply quantization.
+oneshot(model=model, recipe=recipe)
+
+# Save to disk in compressed-tensors format.
+SAVE_DIR = "/raid/engine/dsikka/" + MODEL_ID.rstrip("/").split("/")[-1] + "-FP8-Dynamic"
```
Comment (kylesayrs) [inline examples/quantization_w8a8_fp8/gemma4_example.py]: Change path

File: src/llmcompressor/modeling/gemma4.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        self.num_experts = original.num_experts
+        self.hidden_dim = original.hidden_dim
+        self.intermediate_dim = original.intermediate_dim
+        self.calibrate_all_experts = calibrate_all_experts
+
+        # Unpack the 3D expert weights into individual MLP modules
+        # Register experts directly as numbered children to avoid double nesting
+        # (HF has layers[i].experts, so we want layers[i].experts.0,
+        # not layers[i].experts.experts.0)
+        expert_list = Gemma4TextExpertsList(config.text_config, original)
```
[the remark this answers is not in the document]
Comment (kylesayrs) [reply src/llmcompressor/modeling/gemma4.py:45]: This is true, GemmaForCausalLM uses a text config
https://github.com/huggingface/transformers/blob/main/src/transformers/models/gemma4/modeling_gemma4.py#L1686

File: src/llmcompressor/modeling/gemma4.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        down_data = original.down_proj.data  # [num_experts, hidden, inter]
+
+        for i in range(self.num_experts):
+            gate_up = gate_up_data[i]  # [2*intermediate, hidden]
+            down = down_data[i]  # [hidden, intermediate]
+
+            # gate_up_proj stores [gate; up] stacked along dim 0
+            # nn.Linear weight is [out_features, in_features]
+            self[i].gate_proj.weight.data = (
+                gate_up[:intermediate_size, :].clone().contiguous()
```
Comment (kylesayrs) [inline src/llmcompressor/modeling/gemma4.py:111]: This is an out-of-scope problem, but creating the new `Gemma4TextExpertsList` class while the original class lives causes a memory spike, which may cause oom for large models (for example, kimi-k2 is 34Gb, doubling that to 68Gb is a large cost).

Comment (kylesayrs) [inline src/llmcompressor/modeling/gemma4.py:111]: Clone and contiguous are redundant

File: examples/quantization_w4a4_fp4/gemma4_example.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ds = ds.map(preprocess_function, batched=False, remove_columns=ds.column_names)
+
+
+def data_collator(batch):
+    assert len(batch) == 1
+    return {
+        key: (
+            torch.tensor(value)
+            if key != "pixel_values"
+            else torch.tensor(value, dtype=torch.bfloat16).squeeze(0)
```
Comment (kylesayrs) [inline examples/quantization_w4a4_fp4/gemma4_example.py:70]:
```suggestion
            else torch.tensor(value, dtype=model.dtype).squeeze(0)
```

**Comment by dsikka** [comment]
@coderabbitai review and add labels

**Comment by dsikka** [comment]
@coderabbitai Review and label

**Comment by dsikka** [comment]
@coderabbitai review

**Comment by dsikka** [comment]
@coderabbitai full review
